### DISCUSSION TITLE
Pihole with v2ray

### DISCUSSION THREAD

**Discussion by bart3nder** [v2fly/v2ray-core#2223] [Q&A]
Hello y'all! I have a pihole resolver listening on `127.0.0.15` port `53`. I was wondering how can I configure my config (vmess/vless/trojan protocols) file to use it and pass resolver the domain via pihole? I have used this but it didn't work.
```
  "dns": {
    "servers": [
      "127.0.0.15"
    ]
  },
```

**Comment by mohamadfarahani** [comment]
Hey there do you have any update on this subject?

**Comment by mydogshitgold** [comment]
Domain name resolution takes place on the remote server when using protocols such as HTTP, SOCKS5 as a proxy for popular web browsers. Unless you change the `routing.domainStrategy`, `dns.servers` will not be used.

In addition, the `dns.servers` setting only affects V2Ray's name resolution. This means that your operating system and other programs such as `curl`, `apt`,  `git`, will continue to use the default name server unless you explicitly configure them otherwise.

**Reply by mydogshitgold** [reply]
There are two ways to use the DNS service of the V2Ray client globally, and by extension, your preferred DNS servers.

### Via the dokodemo-door inbound 

Create a dokodemo-door inbound on designated host:port, then route traffics to a DNS outbound. You will still need to explicitly configure programs to use this inbound.

### Transparent proxy

With transparent proxy, you can easily hijack all DNS traffic originated from your local network without any manual configuration. However, this requires the modification of routing table and firewall, and currently only works on Linux.

**Reply by silvermoon1400** [reply]
Hello, how can I talk to you privately? I need your help

**Reply by mydogshitgold** [reply]
Why? You can simply ask here.

**Reply by bart3nder** [reply]
Hi, sorry for very late response. I actually solve my issue by changing the system resolvers in `/etc/resolv.conf`. I understand your solution by I can not deploy it on my server, Could you please provide the answer in code? Like, Is this the correct way to use the **dokodemo-door inbound**? (I have a pihole resolver listening on `127.0.0.15` port `53`)
```
        {
            "tag": "DOKODEMO",
            "listen": "127.0.0.1",
            "port": 12345,
            "protocol": "dokodemo-door",
            "settings": {
                "address": "127.0.0.15",
                "network": "tcp, udp"
            },
```
and routing:
```
    "routing": {
        "domainStrategy": "AsIs",
        "rules": [
            {
                "type": "field",
                "inboundTag": "DOKODEMO",
                "outboundTag": "freedom"
            },
            {
                "type": "field",
                "ip": [
                    "geoip:private"
                ],
                "outboundTag": "blocked"
            }
        ]
```

**Reply by silvermoon1400** [reply]
> چرا؟ به سادگی می توانید اینجا بپرسید.
In Iran, we use veles+tls+cdn(cloudflare) and we use cloudflare's cdn in the address field.
But the problem is that Cloud Flare IPs are constantly being filtered, is there a solution for it?

**Reply by mydogshitgold** [reply]
@bart3nder 

The following is a minimal configuration
```jsonc
{
  "inbounds": [
    {
      "tag": "dns_in",
      "protocol": "dokodemo-door",
      // optinally
      // "listen": "127.0.0.1",
      "port": 10053,
      "settings": {
        // The specified DNS server here is used solely for queries that are not A or AAAA.
        // You need to specified it again in ".dns" configuration
        "address": "127.0.0.15",
        "port": 53,
        "network": "tcp,udp"
      }
    }
  ],
  // To ensure that the built-in DNS module processes your DNS request, you need to use an DNS outbound.
  "outbounds": [
    {
      "tag": "dns",
      "protocol": "dns",
      // Make sure that all DNS queries are sent through the proxy to prevent any queries from being sent without it.
      "proxySettings": {
        // replace this with tag of your proxy outbound
        "tag": "tag_of_your_primary_proxy_outbound"
      }
    }
  ],
  "routing": {
    "rules": [
      {
        "outboundTag": "dns",
        "inboundTag": [
          "dns_in"
        ],
        "type": "field"
      }
    ]
  },
  "dns": {
    "servers": [
      // here, your pihole again, for built-in DNS module
      "127.0.0.15"
    ]
  }
}
```

**Reply by mydogshitgold** [reply]
> In Iran, we use veles+tls+cdn(cloudflare) and we use cloudflare's cdn in the address field.
But the problem is that Cloud Flare IPs are constantly being filtered, is there a solution for it?

Keep in mind that Iran has been reported by net4people to use TLS fingerprinting and SNI-based ban. To confirm whether the ban is in effect, you can check for abnormal RST packets that occur rapidly in succession. To obtain a comprehensive list of IPs used by Cloudflare, you may check [bgp.tool](https://bgp.tools/).

**Reply by nursery01** [reply]
> To obtain a comprehensive list of IPs used by Cloudflare, you may check bgp.tool.

https://www.cloudflare.com/ips

**Reply by mydogshitgold** [reply]
> > To obtain a comprehensive list of IPs used by Cloudflare, you may check bgp.tool.
> 
> https://www.cloudflare.com/ips

It is not an exhaustive list. e.g. No IPs  (BYOIP) from AS13335 are included.

**Comment by bart3nder** [comment]
@mydogshitgold 
Hello again
Bro I tested your minimal sample but I'm getting this:
```
2023/04/10 14:29:24 DNS accepted udp:127.0.0.15:53 [xray.system.b8a60de6-aa28-4e22-887a-21ddb5160d38 -> blocked]
```
What's wrong? btw I didn't quite understand what do you mean by `tag_of_your_primary_proxy_outbound`. what should I write? for example `freedom`?

**Reply by mydogshitgold** [reply]
You should replace it with the tag you give to your outbound proxy. If you have no idea what that is, read the [official documentations](https://www.v2fly.org/config/overview.html),(English version is incomplete, so you should use the Mandarin version with [translation](https://www.deepl.com/translator))

Also, this configuration is for reference only, you will need to expand it.

**Reply by mydogshitgold** [reply]
It also appears that you are using [xray-core](https://github.com/xtls/xray-core), which may behave slightly differently to v2ray-core.

**Reply by nursery01** [reply]
@mydogshitgold ~~Why are you so good to this person only? You only taught him  many times.~~

**Reply by nursery01** [reply]
@mydogshitgold Meowwww？...No, I won't send email to any giter.

**Reply by bart3nder** [reply]
Hello again.
Brother I guess you are talking about implementing this at client side, while I meant how to use Pihole on the v2ray server (server side). If you have any solution for that I would love to hear it as well, otherwise thanks a lot for your time, appreciate it.

**Reply by mydogshitgold** [reply]
This example also applies to the server side (v2ray does not distinguish between server and client conceptually). Just remove the `proxySettings`. You can also make v2ray reset blocklisted connections via IP rules on the server side.

**Reply by mezhgano** [reply]
> @mydogshitgold Hello again Bro I tested your minimal sample but I'm getting this:
> 
> ```
> 2023/04/10 14:29:24 DNS accepted udp:127.0.0.15:53 [xray.system.b8a60de6-aa28-4e22-887a-21ddb5160d38 -> blocked]
> ```
> 
> What's wrong? btw I didn't quite understand what do you mean by `tag_of_your_primary_proxy_outbound`. what should I write? for example `freedom`?


I'm facing exact same problem and looks like it causing by this routing rule:
```json
{
   "type": "field",
   "ip": ["geoip:private"],
     "outboundTag": "blocked"
}
```

DNS requests to private ip just being blocked.

To solve this i had to add new rule above:
```json
{
    "type": "field",
    "ip": ["LOCAL_DNS_SERVER_IP"],
    "port": 53,
    "network": "tcp,udp",
    "outboundTag": "direct"
}
```

After that it start works and i see queries in Pi-hole interface.

I'm completely noob in XRAY configuration, so if i make something wrong, please correct me.

**Comment by silvermoon1400** [comment]
> سلام به همه شما من یک Pihole Resolution دارم که روی `127.0.0.15`پورت گوش می دهد `53`. می خواستم بدانم چگونه می توانم فایل کانفیگ (پروتکل های vmess/vless/trojan) خود را پیکربندی کنم تا از آن استفاده کنم و Resolver دامنه را از طریق pihole ارسال کنم؟ من از این استفاده کردم ولی جواب نداد
> 
> ```
>   "dns": {
>     "servers": [
>       "127.0.0.15"
>     ]
>   },
> ```

Hello, did you finally succeed?
